### ISSUE TITLE
pushenv with empty string causes unexpected behaviour

### ISSUE THREAD

**Issue by d-kirsten** [TACC/Lmod#627]
# Description

The bug report is about two different behaviours around pushenv. I include both here but if you want I can split one of them off into a new bug report.

It's generally about the behaviour of using an empty string with pushenv. The first concerns normal modulefiles while the second is about using pushenv in SitePackage. I think the problem is with treating modulefiles as untrusted while SitePackage runs normally without a pcall. I assume a fix would solve both simultaneously.

# Reproduce
 
## Reproduce with modulefiles

1. Create two modulefiles, called `first.lua` and `second.lua`, with the following content:

`$ cat first.lua`

```lua
pushenv("TEST", "")
```

`$ cat second.lua`

```lua
pushenv("TEST", "second")
```

Then load both modules, first then second, and then unload second. Check the value of `TEST`:

```sh
ml first
ml second
ml -second
env | grep -i test
```
will produce the output
```
TEST=second
__LMOD_STACK_TEST=::
```

## Reproduce with SitePackage

Create two empty `a.lua` and `b.lua` modulefiles.

```sh
touch a.lua
touch b.lua
```

and add the following to `SitePackage.lua`, while making sure it will be used by Lmod

```lua
local hook = require("Hook")

function myload(...)
   local argA = table.pack(...)
   for _, v in ipairs(argA) do
      if v.modFullName == "a" then
         pushenv("TEST", "")

      elseif v.modFullName == "b" then
         pushenv("TEST", "b")
      end
   end
end

hook.register("load", myload)

function myunload(...)
   local argA = table.pack(...)
   for _, v in ipairs(argA) do
      if v.modFullName == "a" then
         pushenv("TEST", "")

      elseif v.modFullName == "b" then
         pushenv("TEST", "b")
      end
   end
end

hook.register("unload", myunload)
```

Load modules `a` and `b`, then unload module `b` again.

```sh
ml a
ml b
ml -b
```

This will cause an error within `decode64`

```
/usr/bin/lua: /home/USER/code/lmod/lmod/libexec/../tools/base64.lua:18: attempt to perform arithmetic on a nil value (local 'value')
stack traceback:
	/home/USER/code/lmod/lmod/libexec/../tools/base64.lua:18: in upvalue 'lsh'
	/home/USER/code/lmod/lmod/libexec/../tools/base64.lua:83: in function 'base64.decode64'
	/home/USER/code/lmod/lmod/libexec/MainControl.lua:463: in function 'MC_Unload.pushenv'
	/home/USER/code/lmod/lmod/libexec/modfuncs.lua:278: in function 'pushenv'
	/home/USER/code/lmod/config/SitePackage.lua:24: in function 'myunload'
	(...tail calls...)
	/home/USER/code/lmod/lmod/libexec/Hub.lua:521: in function 'Hub.unload'
	/home/USER/code/lmod/lmod/libexec/MainControl.lua:1148: in function 'MC_Unload.load'
	/home/USER/code/lmod/lmod/libexec/MainControl.lua:1173: in function 'MC_Unload.unload_usr'
	/home/USER/code/lmod/lmod/libexec/modfuncs.lua:770: in function 'unload_usr_internal'
	/home/USER/code/lmod/lmod/libexec/cmdfuncs.lua:532: in upvalue 'l_usrLoad'
	/home/USER/code/lmod/lmod/libexec/cmdfuncs.lua:575: in function 'Load_Usr'
	/home/USER/code/lmod/lmod/libexec/lmod:513: in function 'main'
	/home/USER/code/lmod/lmod/libexec/lmod:584: in main chunk
	[C]: in ?
```

# Expected behaviour

## Expected with modulefiles

`TEST` should have been the empty string when unloading second.

## Expected with modulefiles

No error should have happened and `TEST` should then be the empty string.

# Desktop (please complete the following information):
 - OS: Linux
 - Linux distribution: Rocky Linux 8.7
 - Lmod Version:
```
Modules based on Lua: Version 8.7.19  2023-01-17 10:45 -06:00
    by Robert McLay mclay@tacc.utexas.edu

Changes from Default Configuration
----------------------------------

Name                         Where Set  Default      Value
----                         ---------  -------      -----
LFS_VERSION                  D          1.6.3        1.8.0
LMOD_PACKAGE_PATH            D          nil          /home/USER/code/lmod/config
LMOD_PAGER                   C          less         /usr/bin/less
LMOD_SYSTEM_DEFAULT_MODULES  D          __unknown__  <empty>
LMOD_TCLSH                   C          tclsh        /usr/bin/tclsh
MODULEPATH_ROOT              E                       /opt/modulefiles
PATH_TO_LUA                  C          lua          /usr/bin/lua


Where Set -> D: default, E: environment, C: configuration
             lmod_cfg: lmod_config.lua SitePkg: SitePackage StdPkg: StandardPackage
             Other: Set somewhere outside of normal locations
```

# Additional context

The following information might be incorrect as I had to use `-D` to slowly step through the code. Feel free to correct me when I'm wrong.

The difference in behaviour between modulefiles and SitePackage is due to modulefiles going through a `pcall`.

In both cases the error happens in `popenv` from `MainControl.lua` https://github.com/TACC/Lmod/blob/328a763d3396162f6b6a102614a488b0c412f263/src/MainControl.lua#L463

because `v64` is actually a single space.  base64 cannot decode it which causes the error when using SitePackage. With modulefiles this is wrapped in a `pcall` which would explain why it doesn't abort and also why `TEST` never gets set to the value of the first module.

Concerning why `v64` is a single space. It gets set by `pop` in `Var.lua`
https://github.com/TACC/Lmod/blob/328a763d3396162f6b6a102614a488b0c412f263/src/MainControl.lua#L458

which should return the second value. That second value is supposed to be the empty string but encoded to base64.

When we loaded the first module which `pushenv` the empty string, we use `pushenv` from `Maincontrol.lua`. Which at some point calls `prepend` in `Var.lua` https://github.com/TACC/Lmod/blob/328a763d3396162f6b6a102614a488b0c412f263/src/MainControl.lua#L426

In `prepend` we call `path2pathA` who gets passed the empty string as `value` https://github.com/TACC/Lmod/blob/328a763d3396162f6b6a102614a488b0c412f263/src/Var.lua#L367

It gets passed the empty string because `encode64("")` returns the empty string. `path2pathA` returns `{ ' ' }`. Which is then used in `prepend` to insert a single space as a key. This is retrieved by `pop` and fails in `decode64`, causing the issue reported by me.

**Comment by rtmclay** [comment]
Thanks for the looking at the code.  It was a help.  I'm pleased that you could track down where the issue was.  

Also your comments about the different behavior between your two cases finally let me understand about an issue that I have had for a long time.  Since Lmod does use a pcall() as part of the sandbox.  Lua sometimes catches the error and sometimes it just keeps going where there is an error.  This has existed in Lua since 5.1 through 5.4.  Maybe now I can put together a test case to the Lua developers.

In any event, I was able to reproduce your issue and fix it for me.  This fix can be found in the testing branch.  Please test either the testing branch or when Lmod 8.7.21 comes out.

**Comment by d-kirsten** [comment]
I tested the testing branch. It did fix the issue with the empty string.

But if I change the empty string to the boolean `false` (in module `first`), then I get an error when I load that module

```
Lmod has detected the following error:  Unable to load module because of error when evaluating modulefile:
     /home/USER/code/lmod/modulefiles/Core/first.lua: /home/USER/code/lmod/lmod/libexec/../tools/base64.lua:52: bad argument #1 to 'len' (string expected, got boolean)
     Please check the modulefile and especially if there is a line number specified in the above message
While processing the following module(s):
    Module fullname  Module Filename
    ---------------  ---------------
    first            /home/USER/code/lmod/modulefiles/Core/first.lua
```

Might this be related to the commit https://github.com/TACC/Lmod/commit/69457d1c6aec9a318107531f243f31ee27fbd2b9

I'm surprised you removed the false check.

**Comment by rtmclay** [comment]
I have converted pushenv("TEST",false) to unset TEST.  Please test this new version on the testing branch.

Note that the only thing that one can do with env vars is to set it to a string or unset it.  So pushenv("FOO",true) and pushenv("FOO","true") are the same.

**Comment by d-kirsten** [comment]
I can confirm that both behaviours work as expected now.

Thank you.

**Comment by rtmclay** [comment]
This fix as been integrated into Lmod 8.7.21
